src/android/Dgram.java

```diff
@@ -75,198 +48,194 @@ public void run() {
         }
 
         public void run() {
             try {
-                if (this.m_datagramSocket.isClosed()) {
+                if (datagramSocket.isClosed()) {
                     Log.d(TAG, "Trying to send but socket closed");
                     return;
                 }
 
                 // Threaded send to prevent NetworkOnMainThreadException
-                final byte[] bytes = this.m_message.getBytes("UTF-8");
+                final byte[] bytes = this.message.getBytes(StandardCharsets.UTF_8);
                 final DatagramPacket packet = new DatagramPacket(
                         bytes,
                         bytes.length,
-                        InetAddress.getByName(this.m_address),
-                        this.m_port
+                        InetAddress.getByName(this.address),
+                        this.port
                 );
-                this.m_datagramSocket.send(packet);
-                Dgram.this.m_javascriptQueue.put("_onSend(" + this.m_datagramSocketId + ");");
-            } catch (InterruptedException e) {
-                Log.d(TAG, e.toString());
+                datagramSocket.send(packet);
+                sendCallback.success();
             } catch (Exception e) {
                 Log.e(TAG, "Send exception: " + e.toString(), e);
-
-                try {
-                    Dgram.this.m_javascriptQueue.put("_onError("
-                            + this.m_datagramSocketId
-                            + ",'send','"
-                            + e.toString() + "');");
-                } catch (Exception innerException) {
-                    Log.e(TAG, "Send exception: " + e.toString());
-                }
+                sendCallback.error("Send exception: " + e.toString());
             }
         }
     }
 
-    private void open(
-            final int datagramSocketId,
-            final int port,
-            final boolean isBroadcast
-    ) throws SocketException {
+    @Override
+    public boolean execute(final String action, final JSONArray data,
+                           final CallbackContext callbackContext) throws JSONException {
+        Log.e(TAG, "Call to execute " + action + " " + data.toString());
+
+        if (!(action.equals(OPEN_ACTION) || action.equals(ON_MESSAGE_ACTION) || action.equals(SEND_ACTION) || action.equals(CLOSE_ACTION))) {
+            // Returning false results in an INVALID_ACTION error,
+            // which translates to an error callback in the JavaScript
+            return false;
+        }
+
+        if (datagramSocket == null && !action.equals(OPEN_ACTION)) {
+            callbackContext.error("DatagramSocket has not been opened!");
+            return true;
+        }
+
+        switch (action) {
+            case OPEN_ACTION:
+                openSocket(action, data, callbackContext);
+                break;
+
+            case ON_MESSAGE_ACTION:
+                startListening(callbackContext);
+                break;
```

Comment:
@gareth-ivatt-fivestars talked about having the concept of opening and listening just be one thing. I'm not sure if that applied to the JS layer or the Java layer... because we wanted separate callbacks for each I think?

Reply:
That might have been after I was given the API spec - let me know if you have a strong preference / know what it is supposed to be and I can update it

Reply:
i think it significantly simplifies things if open and listen are combined.  it would mean the callback for the listener would need to be a parameter to open.  the nice thing is it creates a contract that the socket is open before we start trying to listen for messages (potentially using a null pointer).